### PR TITLE
Modify `makeChainedTransactions` using `TxBuilder`

### PR DESCRIPTION
**Pull request by MichaelKim20**
This utility uses well-known keys (`WK.Keys`) for signature and expect
output created (via `split`, `merge`, etc...) to use them as addresses.

We used this to improve 'makeChainedTransactions'.
We haven't improved makeTransactionForFreezing yet. We're going to continue with this, too.

And we couldn't apply the new signature of the transaction. It will be applied to the next step.

Relates to #907

### CODE REVIEW

**Comment by Geod24** [comment]
The first commit seems rather independent and should be its own PR.

File: source/agora/utils/Test.d [hunk 1]
```diff
@@ -624,62 +627,274 @@ unittest
...
+}
 
-    Returns:
-        A newly created, valid, and signed `Transaction` spending `input`.
+/// Get an `OutputRef` from a `Transaction`
+public inout(OutputRef) oref (inout Transaction tx, uint index)
+{
+    assert(tx.outputs.length > index);
+    return inout(OutputRef)(tx, index);
+}
```
Comment (Geod24) [inline source/agora/utils/Test.d]: This can be removed and replaced by just a constructor call.

File: source/agora/utils/Test.d [hunk 2]
```diff
@@ -624,62 +627,274 @@ unittest
...
+        const txHash = this.data.hashFull();
+        // Sign all inputs using WK keys
+        foreach (idx, ref in_; this.inputs)
+        {
+            auto ownerKP = getOwnerKeyPair(in_.output.address, unknowns);
+            assert(ownerKP !is KeyPair.init,
+                    "Address not found in Well-Known keypairs: "
+                    ~ in_.output.address.toString());
+            auto signature = () @trusted { return ownerKP.secret.sign(txHash[]); }();
+            this.data.inputs[idx].signature = signature;
```
Comment (Geod24) [inline source/agora/utils/Test.d]: Merge those two statements

File: source/agora/utils/Test.d [hunk 3]
```diff
@@ -624,62 +627,274 @@ unittest
...
-    const resultHash = result.hashFull();
-    foreach (idx, ref in_; result.inputs)
+    private KeyPair getOwnerKeyPair (scope const PublicKey address,
+        scope const(KeyPair)[] unknowns = null) @safe
     {
-        auto rng = from.find!(a => a.address == input.outputs[idx].address);
-        assert(rng.length);
-        const owner = rng[0];
-        in_.signature = sign(owner, resultHash);
+        version(unittest)
```
Comment (Geod24) [inline source/agora/utils/Test.d]: This whole module is made to be used from tests so having this special case is a bit surprising

Comment (Geod24) [reply source/agora/utils/Test.d]: I see why it was introduced but I think there are better ways. Let's discuss a bit later.

File: source/agora/utils/Test.d [hunk 4]
```diff
@@ -624,62 +627,274 @@ unittest
...
+
+    /***************************************************************************
+
+        Attaches one output of a transaction
+
+    ***************************************************************************/
+
+    public void attach (scope const Transaction tx, uint index) @trusted
+    {
+        this.inputs ~= cast(OutputRef)oref(tx, index);
```
Comment (Geod24) [inline source/agora/utils/Test.d]: Why the `cast` ?

File: source/agora/utils/Test.d [hunk 5]
```diff
@@ -624,62 +627,274 @@ unittest
...
+
+    /***************************************************************************
+
+        Attaches all outputs of a transaction
+
+    ***************************************************************************/
+
+    public void attach (scope const Transaction tx) @trusted
+    {
+        this.inputs ~= cast(OutputRef[])oref(tx).array;
```
Comment (Geod24) [inline source/agora/utils/Test.d]: Why the `cast` ?

File: source/agora/utils/Test.d [hunk 6]
```diff
@@ -624,62 +627,274 @@ unittest
...
+            unknowns = unknown `KeyPair`s.
+                finds in `unknowns` first. If not, finds it in 'WK'.
+
+        Returns:
+            returns the final built `Transaction`.
+
+    ***************************************************************************/
+
+    public Transaction sign (TxType type = TxType.Payment,
+        scope const(KeyPair)[] unknowns = null) @safe
```
Comment (Geod24) [inline source/agora/utils/Test.d]: Why is this needed ?

File: source/agora/utils/Test.d [hunk 7]
```diff
@@ -624,62 +627,274 @@ unittest
...
+        Params:
+            share = amount to be distributed to one address
+            use_remainder = whether to use the remaining amount
+                after distribution.
+            toward = address of destination to be transferred.
+
+    ***************************************************************************/
+
+    public void distribute (scope const Amount share, bool use_remainder,
+        scope const PublicKey[] toward...) @safe
```
Comment (Geod24) [inline source/agora/utils/Test.d]: This offers a rather confusing API:
- In general, we want to offer a simple API that covers most needs. Users can always manually modify the tx being built in order to satisfy their specific needs;
- As a result, `use_remainder` is the first thing that stands out. Remove it;
- Likewise, why is `share` needed ? The idea is that if you want to build complex scenarii, you will chain calls to `merge` and `split`.

File: source/agora/utils/Test.d [hunk 8]
```diff
@@ -624,62 +627,274 @@ unittest
...
+}
+
+/// Test of TxBuilder.getOwnerKeyPair
+unittest
+{
+    TxBuilder builder;
+    KeyPair unknown1 = KeyPair.fromSeed(
+        Seed.fromString("SCLEH2TOYRNMH6NS2POFFMINKY6DYBWKIIW7A7BAZ3H5BE7KZIWUKF42"));
+    KeyPair unknown2 = KeyPair.fromSeed(
+        Seed.fromString("SDDQOCBBREPZZJDUWMZUGEUZVNTYBCF4AJPAQYIHU6TINNV7SXMKBUAJ"));
```
Comment (Geod24) [inline source/agora/utils/Test.d]: Why not `random` ?

File: source/agora/utils/Test.d [hunk 9]
```diff
@@ -624,62 +627,274 @@ unittest
...
+    /***************************************************************************
+
+        Splits the attached input into multiple outputs.
+
+        Params:
+            toward = address of destination to be transferred.
+
+    ***************************************************************************/
+
+    public void split (scope const PublicKey[] toward...) @safe
```
Comment (Geod24) [inline source/agora/utils/Test.d]:
```suggestion
    public ref typeof(this) split (scope const PublicKey[] toward...) @safe return
```

This will allow us to do `auto tx = TxBuilder(...).split(...).merge(...).split(...).sign();`

**Review by Geod24 — COMMENTED** [review]
Haven't reviewed the last commit yet.

File: source/agora/utils/Test.d (hunk 5 above)
Comment (MichaelKim20) [reply source/agora/utils/Test.d]: The following message came out.
`cannot append type const(OutputRef) to type OutputRef[]`

File: source/agora/utils/Test.d (hunk 6 above)
Comment (MichaelKim20) [reply source/agora/utils/Test.d]: The parameters unknowns are added for `system integration test` and
and `source/tests/unit/*.d`.

File: source/agora/utils/Test.d (hunk 7 above)
Comment (MichaelKim20) [reply source/agora/utils/Test.d]: I will remove

File: source/agora/utils/Test.d (hunk 6 above)
Comment (MichaelKim20) [reply source/agora/utils/Test.d]: They need a genesis KeyPair of coin-net.

File: source/agora/utils/Test.d (hunk 8 above)
Comment (MichaelKim20) [reply source/agora/utils/Test.d]: That's a good idea.

File: source/agora/utils/Test.d (hunk 5 above)
Comment (MichaelKim20) [reply source/agora/utils/Test.d]: I'll change the member like this and remove `cast`.
``` D
private const(OutputRef)[] inputs;
```

**Comment by Geod24** [comment]
Could you raise a PR with the first commit only ?

File: source/agora/utils/Test.d [hunk 10]
```diff
@@ -624,62 +627,309 @@ unittest
...
+    /// The index of the `Output` being referenced
+    public uint        index;
+}
+
+/// Get a range of all `OutputRef` from a `Transaction`
+public auto oref (const Transaction tx)
+{
+    import std.range;
+    return iota(tx.outputs.length)
+        .map!(idx => OutputRef(cast(Transaction)tx, cast(uint)idx));
```
Comment (Geod24) [inline source/agora/utils/Test.d]: This is casting away `const` and is a big no-no.

File: source/agora/utils/Test.d [hunk 11]
```diff
@@ -624,62 +627,309 @@ unittest
...
+/// A reference to an output
+/// It has a `Transaction` and an index of `Outputs` that it references.
+public struct OutputRef
+{
+    this (Transaction tx, uint index)
+    {
+        assert(tx.outputs.length > index);
+        this.tx = tx;
+        this.index = index;
+    }
```
Comment (Geod24) [inline source/agora/utils/Test.d]: Let's just remove it. The only advantage of this constructor is to do bounds checking, but it will blow up when accessing `output` anyway. And if we wanted to keep it, we would need to add the attributes (`@safe pure nothrow @nogc`).

File: source/agora/utils/Test.d [hunk 12]
```diff
@@ -624,62 +627,309 @@ unittest
...
+        // Finalize the transaction by adding inputs
+        foreach (ref in_; this.inputs)
+            this.data.inputs ~= Input(in_.hash(), in_.index);
+
+        // Get the hash to sign
+        const txHash = this.data.hashFull();
+        // Sign all inputs using WK keys
+        foreach (idx, ref in_; this.inputs)
+        {
+            auto ownerKP = getOwnerKeyPair(in_.output.address);
```
Comment (Geod24) [inline source/agora/utils/Test.d]: Missing `this.`

File: source/agora/utils/Test.d [hunk 13]
```diff
@@ -624,62 +627,309 @@ unittest
...
+    ***************************************************************************/
+
+    public ref typeof(this) merge (scope const PublicKey toward)
+        @safe return
+    {
+        if (this.inputs.length == 0)
+            return this;
+
+        Amount amount;
+        if (this.inputs.all!(val => amount.add(val.output.value)))
```
Comment (Geod24) [inline source/agora/utils/Test.d]: Since those are test utilities, we shouldn't ignore failures like this.

File: source/agora/utils/Test.d [hunk 14]
```diff
@@ -624,62 +627,309 @@ unittest
...
+
+        Params:
+            tx = transaction
+
+        Returns:
+            reference of this
+
+    ***************************************************************************/
+
+    public ref typeof(this) attach (scope const Transaction tx) @trusted return
```
Comment (Geod24) [inline source/agora/utils/Test.d]: Why `scope` ? Why `@trusted` ?

File: source/agora/utils/Test.d [hunk 15]
```diff
@@ -624,62 +627,309 @@ unittest
...
+
+        Returns:
+            reference of this
+
+    ***************************************************************************/
+
+    public ref typeof(this) attach (scope const Transaction tx, uint index)
+        @trusted return
+    {
+        this.inputs ~= OutputRef(cast(Transaction)tx, index);
```
Comment (Geod24) [inline source/agora/utils/Test.d]: We should almost never need to `cast` away `const`.

**Review by Geod24 — CHANGES_REQUESTED** [review]
Haven't finished reviewing, but looks like it's going in the right direction.
There is a misunderstanding for the transaction builder though, it should modify its current state, so `attach` should just append an identity transformation and `split` and `merge` should act on the `output`.

**Comment by Geod24** [comment]
Rebased on master and applied a few of the improvement, will apply the `output` bit soon.

File: source/agora/utils/Test.d [hunk 16]
```diff
@@ -146,89 +152,110 @@ unittest
...
+    else
+    {
+        auto builders = iota(Block.TxsInBlock).map!(_ => new TxBuilder(unknowns)).array;
+        foreach (block_idx; 0 .. block_count)
         {
-            input = Input(hashFull(prev_txs[idx % Block.TxsInBlock]), 0);
-        }
+            uint[] offsets;
+
+            iota(Block.TxsInBlock).each!(_ => offsets ~= 0);
```
Comment (Geod24) [inline source/agora/utils/Test.d]: That is a very complicated way to write:
```suggestion
            uint[] offsets = new uint[](Block.TxsInBlock);
```

File: source/agora/utils/Test.d [hunk 17]
```diff
@@ -146,89 +152,110 @@ unittest
...
+
+            int getAvableIndex (uint last)
```
Comment (Geod24) [inline source/agora/utils/Test.d:220]:
```suggestion
            int getAvailableIndex (uint last)
```

I suppose that's what you meant ?

**Comment by Geod24** [comment]
@MichaelKim20 : I have updated this with my WIP. I removed the `amount` function and added `draw(amount, PublicKey[]...)` instead, as `amount` was mostly used after `split`.
I also changed the initial implementation to use a `refund` transaction instead of always iterating on the input, so we can do things like `TxBuilder(addr).split(...).attach(...).split(...)`, etc...
However, the `split` implementation hasn't been fixed yet (will do that soon).
There is only one test which doesn't pass, because now we have more diversity in keys and we're not always using the genesis keypair. I will fix it soon too.

**Comment by Geod24** [comment]
Re-updated. Changes:
- Renamed `GenesisMainTransactions` to `genesisSpendable`. Made it return a range the filters the transactions for future-proofness;
- Added an example on how one could replace `makeChainedTransactions` in `source/agora/common/TransactionPool.d`. After much thought I will just the content of `makeChainedTransaction` again, although that might take a bit of time.

**Comment by Geod24** [comment]
Just rebased on top of #965 after adding range interface and there was only one usage of the VLA, yay

**Comment by Geod24** [comment]
Rebased on v0.x.x, off to the next round of splitting off

File: source/agora/utils/Test.d [hunk 18]
```diff
@@ -148,83 +148,104 @@ unittest
     spends the entire sum of each provided transaction's output as
     set in the parameters.
 
-    If prev_txs is null, the first set of transactions that fill a block will
-    spend the genesis transaction's outputs.
 
     Params:
+        toward = address of destination to be transferred.
```
Comment (AndrejMitrovic) [inline source/agora/utils/Test.d:153]:
```suggestion
        toward = addresses which will receive the payments
```

File: source/agora/utils/Test.d [hunk 19]
```diff
@@ -148,83 +148,104 @@ unittest
...
-        key_pair = the key pair used to sign transactions and to send
-                   the output to
         block_count = the number of blocks that will be created if the
-                      returned transactions are added to the ledger
-        spend_amount = the total amount to spend (evenly distributed)
-        gen_tx = the genesis transaction to refer to for the first set of
-                 transactions. If none set, the one returned by
-                 GenesisTransaction() is used.
+            returned transactions are added to the ledger
+        share = amount to be distributed to one address
```
Comment (AndrejMitrovic) [inline source/agora/utils/Test.d:157]:
```suggestion
        share = amount to be distributed to each address
```

File: source/agora/utils/Test.d [hunk 20]
```diff
@@ -148,83 +148,104 @@ unittest
...
+        use_remainder = whether to use the remaining amount after distribution.
```
Comment (AndrejMitrovic) [inline source/agora/utils/Test.d:158]: What does "to use" mean? Use how?

File: source/agora/utils/Test.d [hunk 21]
```diff
@@ -148,83 +148,104 @@ unittest
...
+        type = type of `Transaction`
+        unknowns = unknown `KeyPair`s.
+            finds in `unknowns` first. If not, finds it in 'WK'.
```
Comment (AndrejMitrovic) [inline source/agora/utils/Test.d:161]: ??

File: source/agora/utils/Test.d [hunk 22]
```diff
@@ -148,83 +148,104 @@ unittest
...
 *******************************************************************************/
 
-public Transaction[] makeChainedTransactions (KeyPair key_pair,
-    const(Transaction)[] prev_txs, size_t block_count,
-    ulong spend_amount = 40_000_000, in Transaction gen_tx = GenesisTransaction)
-    @safe
+public Transaction[] makeChainedTransactions (TxRange) (
+    scope const(PublicKey)[] toward,
+    TxRange prev_txs_range,
+    scope const size_t block_count,
```
Comment (AndrejMitrovic) [inline source/agora/utils/Test.d:171]:
```suggestion
    const size_t block_count,
```

`scope` on a `size_t` does nothing.

File: source/agora/utils/Test.d [hunk 23]
```diff
@@ -148,83 +148,104 @@ unittest
...
+    scope const Amount share = Amount.MinFreezeAmount,
```
Comment (AndrejMitrovic) [inline source/agora/utils/Test.d:172]:
```suggestion
   const Amount share = Amount.MinFreezeAmount,
```

`scope` here is unnecessary.

File: source/agora/utils/Test.d (hunk 17 above)
Comment (AndrejMitrovic) [reply source/agora/utils/Test.d:220]: Why did you mark this comment as resolved but didn't fix the typo?

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]
I'm not sure about this code. Somehow `makeChainedTransactions` is even more complicated than it was before. I thought the idea was that we would make code easier to understand?

**Comment by AndrejMitrovic** [comment]
I've tried to simplify it a bit. There's still more work that should be done however.

```D
public Transaction[] makeChainedTransactions (TxRange) (
    scope const(PublicKey)[] toward,
    TxRange prev_txs_range,
    scope const size_t block_count,
    scope const Amount share = Amount.MinFreezeAmount,
    TxType type = TxType.Payment) @safe
    if (isInputRange!(TxRange))
{
    import agora.consensus.data.Block;

    // https://issues.dlang.org/show_bug.cgi?id=20937
    const(Transaction)[] prev_txs = () @trusted { return prev_txs_range.array; }();

    // sanity checks
    assert(prev_txs.length > 0 && prev_txs.length <= Block.TxsInBlock);
    assert(
        (prev_txs.length == Block.TxsInBlock) ||
        (
            (prev_txs.length < Block.TxsInBlock) &&
            (prev_txs.map!(tx => tx.outputs.length).sum() >= Block.TxsInBlock)
        )
    );

    Transaction[] transactions;

    if (prev_txs.length == Block.TxsInBlock)
    {
        foreach (block_idx; 0 .. block_count)
        {
            foreach (tx_idx; 0 .. Block.TxsInBlock)
                transactions ~= TxBuilder(prev_txs[tx_idx])
                    .draw(share, toward).sign(type);

            prev_txs = transactions[$ - Block.TxsInBlock .. $];
        }
    }
    else
    {
        foreach (block_idx; 0 .. block_count)
        {
            TxBuilder[] builders;
            uint[Block.TxsInBlock] offsets;
            int getAvailableIndex (uint last)
            {
                foreach_reverse (idx; 0 .. last+1)
                    if ((idx < prev_txs.length) && (offsets[idx] < prev_txs[idx].outputs.length))
                        return idx;
                return -1;
            }

            // attach
            uint loop = 0;
            while (true)
            {
                auto tx_idx = loop++ % Block.TxsInBlock;
                auto available_idx = getAvailableIndex(tx_idx);
                if (available_idx < 0)
                    break;

                if (builders.length <= tx_idx)
                    builders ~= TxBuilder(prev_txs[available_idx], offsets[available_idx]);
                else
                    builders[tx_idx].attach(prev_txs[available_idx], offsets[available_idx]);

                offsets[available_idx]++;
            }

            assert(builders.length == Block.TxsInBlock);
            builders.each!(b => transactions ~= b.split(toward).sign(type));

            // refer to tx'es which will be in the previous block
            prev_txs = transactions[$ - Block.TxsInBlock .. $];
        }
    }

    return transactions;
}
```

Tests still pass. 

I still have to reverse-engineer the `getAvailableIndex` stuff.

**Comment by Geod24** [comment]
This has been split in quite a few commits and changes, and the core that was blocking other issues have been resolved, so let's close this. Thanks a lot for your work @MichaelKim20 !
